module/VuFind/src/VuFind/Config/Feature/SecretTrait.php

```
@@ -0,0 +1,70 @@
+ * it under the terms of the GNU General Public License version 2,
+ * as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * @category VuFind
+ * @package  Config
+ * @author   Robby ROUDON <roudonro@msu.edu>
+ * @license  http://opensource.org/licenses/gpl-2.0.php GNU General Public License
+ * @link     https://vufind.org Main Page
+ */
+
+namespace VuFind\Config\Feature;
+
+use Laminas\Config\Config;
+
+/**
+ * Trait to import secret from file rather than a hardcoded config
+ *
+ * @category VuFind
+ * @package  Config
+ * @author   Robby ROUDON <roudonro@msu.edu>
+ * @license  http://opensource.org/licenses/gpl-2.0.php GNU General Public License
+ * @link     https://vufind.org Main Page
+ */
+trait SecretTrait
+{
+    /**
+     * Load a secret value from the specified configuration and key.
+     * Will look for a _file-suffixed version of the key first,
+     * and load the data from a separate file if configured to do so.
+     *
+     * @param Config|array|null $config The config to read from
+     * @param string            $key    The key to retrieve
+     *
+     * @return string|null
+     */
+    protected function getSecretFromConfig(Config|array|null $config, string $key): ?string
+    {
+        if ($config === null) {
+            return null;
+        }
+        if ($config instanceof Config) {
+            $config = $config->toArray();
+        }
+        if (isset($config[$key . '_file']) && $value = file_get_contents($config[$key . '_file'])) {
```

This will still blow up without proper error message if the file doesn't exist. You could get a PHP warning in the log, but without call stack it would be really difficult to find out what configuration was being read since the error would always be emitted from this line.

---

I modified it, do you prefer it the new way? I'm open to suggestions

---

Looks like an improvement to me. I suppose you could add a further `else` clause to use the `error_log` function if `logWarning` is undefined, so it never silently fails.

---

Done